src/app/util/attribute-storage.cpp

```diff
@@ -606,9 +695,20 @@ Status emAfReadOrWriteAttribute(const EmberAfAttributeSearchRecord * attRecord,
                             }
 
                             {
+                                // attribute storage can be
+                                // - singleton: statically allocated in singletonAttributeData global
+                                // - static endpoint: statically allocated in attributeData global
+                                // - dynamic endpoint with dynamic storage: in memory block provided at endpoint instantiation
                                 uint8_t * attributeLocation =
-                                    (am->mask & ATTRIBUTE_MASK_SINGLETON ? singletonAttributeLocation(am)
-                                                                         : attributeData + attributeOffsetIndex);
+                                    (am->mask & ATTRIBUTE_MASK_SINGLETON
```

Comment:
This became very complex logic in nested ternary with ifdefs. Could we split this out into a static function (even if inline) or some if/else instead?

Reply:
done